### PR TITLE
(feat): allow `join=outer` for `concat_on_disk`

### LINKED ISSUE #1503: `concat_on_disk` outer join
### Please describe your wishes and possible alternatives to achieve the desired result.

There should be the option to outer join via `concat_on_disk` and it seems from looking at the original PR that there shouldn't be a problem doing this: https://github.com/scverse/anndata/pull/955

Someone mentioned that `concat` was taking too much memory so it seems better to fix this than investigate that.

### PR DESCRIPTION
**Pull request by ilan-gold**
- [x] Closes #1503
- [x] Tests added
- [x] Release note added (or unnecessary)

### CODE REVIEW

File: src/anndata/experimental/merge.py [hunk 1]
```diff
@@ -536,8 +536,6 @@ def concat_on_disk(
     # Argument normalization
     if pairwise:
         raise NotImplementedError("pairwise concatenation not yet implemented")
-    if join != "inner":
-        raise NotImplementedError("only inner join is currently supported")
```
Comment (ilan-gold) [inline src/anndata/experimental/merge.py:540]: I am trying to become more familiar with the code, but it's shocking this was all that was needed.  That being said, the tests do pass and I looked at the results so 🤷

Comment (ilan-gold) [reply src/anndata/experimental/merge.py:540]: I mean, looking at the code path, it makes some sense.  `concat_on_disk` constructs sensible-looking reindexers for the `alt_dim` axis and then passes those on so there shouldn't be a need for the `join` argument anywhere else despite the option to use it

**Comment by zimea** [comment]
In a first run, the object didn't contain the "var" dimension. Could you check whether you can replicate it?

**Comment by ilan-gold** [comment]
@zimea #1505 is what you want but there are clearly some problems there.  So if you need you `var` you can do something like this for each anndata:
```python
my_adatas = []
for path in my_paths:
    adata = ad.read_h5ad(path)
    del adata.X
    my_adatas += [adata]
merged_in_memory = concat(my_adatas, merge="...")
merged_on_disk.var = merged_in_memory.var
```
which should be manageable unless your `obsm` is also very large, in which case I would `del` that as well.  This isn't pretty but until I also fix #1505, it should work.  Also you will need to choose a `merge` strategy based on what you want: https://anndata.readthedocs.io/en/stable/generated/anndata.concat.html

**Comment by ilan-gold** [comment]
Ok @ivirshup I think I have a slightly better sense of what you mean when you talk about `concat_on_disk` + `dask`.  It seems like `concat_on_disk` won't disappear but will be vastly simplified by using `dask` for which we need #1469.

For example, I have been looking into adding `merge` support and the main stumbling block there appears to be that doing anything with `SparseDataset` classes is difficult i.e., reindexing + writing, which is why https://github.com/scverse/anndata/blob/105f35441482565bc74168642567c8904ca28ccc/src/anndata/experimental/merge.py#L60-L90 exists. So I think to move forward with the various `merge` types we should merge #1469 and then integrate `read_dask_as_elem` here.  Sound good?

**Review by selmanozleyen — APPROVED** [review]
Looks good. Happy to see this was somewhat smooth.

File: tests/test_concatenate_disk.py [hunk 2]
```diff
@@ -166,11 +175,20 @@ def test_anndatas_with_reindex(
...
+        if axis == 1:
+            a.obs_names = [
+                f"{i}-{name}" if obs_ind % 2 else name
+                for obs_ind, name in enumerate(a.obs_names)
+            ]
+        else:
+            a.var_names = [
+                f"{i}-{name}" if var_ind % 2 else name
+                for var_ind, name in enumerate(a.var_names)
+            ]
```
[the remark this answers is not in the document]
Comment (ilan-gold) [reply tests/test_concatenate_disk.py:191]: > What’s this supposed to do? a, 1-a, b, 3-b, c, 4-c?

Yes

> why those numbers?

If it's `2` you're referring to, it was arbitrary. 

The suggestion looks good, will do.  But you're ok with the concept here?

[the remark this answers is not in the document]
Comment (flying-sheep) [reply tests/test_concatenate_disk.py:186]: The tests need to create `AnnData` objects that have partially overlapping names in the off-axis, and then check if joining them works without error and the resulting file can be read and
- when outer-joined has all existing off-axis names and all data that went in
- when inner-joined has the off-axis names shared by all objects and their data

If I understand correctly, this achieves that.

**Review by flying-sheep — APPROVED** [review]
Looks good. It tests correctness, and if performance turns out to be a problem, we can do that later

**Pull request merged**